Add search

albertleinsoo opened the pull request:
Поиск фильмов по названию и режиссёрам

albertleinsoo committed:
fix: поправлен поиск по лайкам и режисcёрам

alex-blag commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -81,5 +81,9 @@ public List<Film> getDirectorFilmsSortedBy(@PathVariable final long directorId,
     }
 
+    @GetMapping("/films/search")
+    public List<Film> searchFilmsByTitleDirector(@RequestParam String query, @RequestParam String by) {
```
Можешь попробовать вместо "String by" использовать List или Set, будет удобнее.

alex-blag commented on src/main/java/ru/yandex/practicum/filmorate/exeptions/FilmSearchByTitleDirectorException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.yandex.practicum.filmorate.exeptions;
+
+public class FilmSearchByTitleDirectorException extends RuntimeException {
```
Так а ошибка-то в чем?

albertleinsoo replied:
Передаётся в сообщении

alex-blag replied:
И называние эксепшена должно указывать ошибку :-) Иначе мог бы назвать просто Exception

albertleinsoo replied:
Название ошибки поиска стало более говорящим

alex-blag commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -115,6 +116,15 @@ public List<Film> getDirectorFilmsSortedBy(long directorId, String sortBy) {
     }
 
+    public List<Film> searchFilmsByTitleDirector(String query, String by) {
+        if (by.equals("title") || by.equals("director") || by.equals("title,director") || by.equals("director,title")) {
```
Если будет не "String by", а List или Set проверять будет удобнее.

albertleinsoo replied:
Согласен, исправил

albertleinsoo replied:
Согласен, поправил

alex-blag commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -437,6 +422,37 @@ public List<Film> getDirectorFilmsSortedBy(long directorId, String sortBy) {
+                        "WHERE UPPER(CONCAT(" +
+                                                (by.contains("title") ? "F.NAME" : "NULL") + ", " +
+                                                (by.contains("director") ? "D.NAME" : "NULL") +
+                                            ")) LIKE UPPER('%" + query + "%')" +
```
А попробовал?

String sqlParam = "%" + query + "%";
LIKE UPPER(?)

albertleinsoo replied:
Не понимаю различия между этими конструкциями. Что так я делаю SQL инъекцию вставляя String query, что String sqlParam = "%" + query + "%";

alex-blag replied:
Полагаю, что тут <LIKE UPPER('%" + query + "%')"> -- возможна инъекция.
А тут <LIKE UPPER(?)> -- нет.

albertleinsoo replied:
Параметр запроса изменён на:
String sqlParam = "'%" + query + "%'";

albertleinsoo committed:
fix: параметр запроса поиска 'by' преобразуется в Set.
Название ошибки поиска стало более говорящим.
Причесал код (Файлы оканчиваются пустой строкой, оптимизированы импорты, поправлены отступы в некоторых местах)

alex-blag commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -24,6 +26,8 @@ public class FilmDbStorage implements FilmStorage {
 
     private final JdbcTemplate jdbcTemplate;
 
+    //private final NamedParameterJdbcTemplate = new NamedParameterJdbcTemplate(jdbcTemplate);
```
-

albertleinsoo replied:
.

albertleinsoo committed:
fix: убран закомментированный код
